miio/gateway/gateway.py

```diff
@@ -0,0 +1,335 @@
+"""Xiaomi Gateway implementation using Miio protecol."""
+
+import logging
+import os
+import sys
+
+import click
+
+import yaml
+
+from ..click_common import command
+from ..device import Device
+from ..exceptions import DeviceError, DeviceException
+
+_LOGGER = logging.getLogger(__name__)
+
+GATEWAY_MODEL_CHINA = "lumi.gateway.v3"
+GATEWAY_MODEL_EU = "lumi.gateway.mieu01"
+GATEWAY_MODEL_ZIG3 = "lumi.gateway.mgl03"
+GATEWAY_MODEL_AQARA = "lumi.gateway.aqhm01"
+GATEWAY_MODEL_AC_V1 = "lumi.acpartner.v1"
+GATEWAY_MODEL_AC_V2 = "lumi.acpartner.v2"
+GATEWAY_MODEL_AC_V3 = "lumi.acpartner.v3"
+
+
+class GatewayException(DeviceException):
+    """Exception for the Xioami Gateway communication."""
+
+
+class Gateway(Device):
+    """Main class representing the Xiaomi Gateway.
+
+    Use the given property getters to access specific functionalities such
+    as `alarm` (for alarm controls) or `light` (for lights).
+
+    Commands whose functionality or parameters are unknown,
+    feel free to implement!
+    * toggle_device
+    * toggle_plug
+    * remove_all_bind
+    * list_bind [0]
+    * bind_page
+    * bind
+    * remove_bind
+
+    * self.get_prop("used_for_public") # Return the 'used_for_public' status, return value: [0] or [1], probably this has to do with developer mode.
+    * self.set_prop("used_for_public", state) # Set the 'used_for_public' state, value: 0 or 1, probably this has to do with developer mode.
+
+    * welcome
+    * set_curtain_level
+
+    * get_corridor_on_time
+    * set_corridor_light ["off"]
+    * get_corridor_light -> "on"
+
+    * set_default_sound
+    * set_doorbell_push, get_doorbell_push ["off"]
+    * set_doorbell_volume [100], get_doorbell_volume
+    * set_gateway_volume, get_gateway_volume
+    * set_clock_volume
+    * set_clock
+    * get_sys_data
+    * update_neighbor_token [{"did":x, "token":x, "ip":x}]
+
+    ## property getters
+    * ctrl_device_prop
+    * get_device_prop_exp [[sid, list, of, properties]]
+
+    ## scene
+    * get_lumi_bind ["scene", <page number>] for rooms/devices"""
+
+    def __init__(
+        self,
+        ip: str = None,
+        token: str = None,
+        start_id: int = 0,
+        debug: int = 0,
+        lazy_discover: bool = True,
+    ) -> None:
+        super().__init__(ip, token, start_id, debug, lazy_discover)
+
+        from . import (
+            Alarm,
+            Radio,
+            Zigbee,
+            Light,
+        )
+
+        self._alarm = Alarm(parent=self)
+        self._radio = Radio(parent=self)
+        self._zigbee = Zigbee(parent=self)
+        self._light = Light(parent=self)
+        self._devices = {}
+        self._info = None
+        self._subdevice_model_map = None
+
+    def _get_subdevice_model_map(self):
```

Comment:
Instead of having to call this prior to accessing `self._subdevice_model_map`, how about defining a property `subdevice_model_map` which does the initialization during the first access?

Something like:
```
@property
subdevice_model_map():
    if self._subdevice_model_map is None:
        # initialize 
    return self._subdevice_model_map
```

That would simplify all the places you need this information :-)

Reply:
Greath suggestion, I already made a commit for this.
Thanks!